### PR TITLE
Add smoldot rpc provider

### PR DESCRIPTION
**Pull request by josepot**
:wave: @jacogr !

Now that the `@substrate/connect` API has a mature and stable API, we think that it would be best to have the PolkadotJs RPC Provider for smoldot exported along side with the other 2 `polkadot-js` RPC providers. This would allow us to keep `@substrate/connect` properly decoupled from `polkadot-js`, which should be the best outcome for both libraries. 

It goes without saying that we (the @substrate/connect team) will be more than happy to help with the maintenance of this new RPC Provider. Also, we would be like to keep in an ongoing communication to ensure that we properly cover (and over time hopefully reduce) the impedance mismatch between the polkadot-js RPC Provider Interface and @substrate/connect.

### CODE REVIEW

File: yarn.lock
Comment (jacogr) [inline yarn.lock:2539]: This bit seems problematic? It needs to be kept in alignment?

**Comment by tomaka** [comment]
Just want to point out that this doesn't use "smoldot", as the documentation and naming implies, but substrate-connect. It is important for users to be aware of the fact that this might rely on the substrate-connect browser extension if it is available.

**Comment by jacogr** [comment]
Thank you. Will take a peek, just scrolled through above.

File: yarn.lock
Comment (josepot) [reply yarn.lock:2539]: yeah! we are in a bit of a chicken and egg situation here... Once `@substrate/connect` stops exporting the polkadot-js rpc provider, then that dependency will stop being there. Although, if you have closer look at this PR, then you will notice that the parts of the API that are currently being used don't rely on `@polkadot/rpc-provider`.

**Comment by josepot** [comment]
> Just want to point out that this doesn't use "smoldot", as the documentation and naming implies, but substrate-connect. It is important for users to be aware of the fact that this might rely on the substrate-connect browser extension if it is available.

Good point! Sorry, I will update it accordingly. I mean, it does use "smoldot" through `@substrate/connect`, but yeah, I will make sure that it's clearer.

**Review by jacogr — COMMENTED** [review]
I like it since this could drive more use, which is brilliant. Made some small comments. In general didn't inspect the provider in-depth since I know it works as expected, just want alignment with the rest of the code.

I do have one concern that this new `createScClient` does not fit with the APIs here at all. So would really like an `new ScProvider(<chain info>)` solution so it fits in with the, erm, API style of the provider interfaces available if it is to be exposed here.

Obviously also want the little dependency catch-22 addressed as soon as this is in, since I really don't want old packages via dependencies to sneak in from these. (If all else fails a `peerDep` of `*` could do the trick, although I have not thought that through, so may have other issues.)

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:15]: >  It basically clobbers the rules for the repo which is there to ensure we don't run into unknowns.

We do have a good set of tests for ensuring that we don't run into those unknowns.

Also, what it tends to happen when enforcing these rules is that the code ends up being worse just for the purpose of shutting up the rules e.g (`.catch(console.log)`), etc. We use to have most of these rules on `substrate-connect` and we got rid of them because they were creating more issues than the issues that they were trying to solve.

> There are obviously places where this needs to apply despite best efforts (false positives, places where ordering is important, e.g. definitions, etc.), but generally we try to keep things consistent.

If you really care that much about having the keys sorted alphabetically, I will update the code accordingly... sure :shrug:

**Comment by josepot** [comment]
> I like it since this could drive more use, which is brilliant. Made some small comments. In general didn't inspect the provider in-depth since I know it works as expected, just want alignment with the rest of the code.

:+1: 

> I do have one concern that this new `createScClient` does not fit with the APIs here at all. So would really like an `new ScProvider(<chain info>)` solution so it fits in with the, erm, API style of the provider interfaces available if it is to be exposed here.

Yeah... This is something that has to do with the impedance mismatch that I mentioned before between the RpcProvider interface of PolkadotJs and the API of `@substrate/connect`. There are an important number of reasons why we decided to go with this API. Perhaps @tomaka is the best person for explaining why an API like `new ScProvider(<chain info>)` is not the best idea for a `@substrate/connect` Provider...

> Obviously also want the little dependency catch-22 addressed as soon as this is in, since I really don't want old packages via dependencies to sneak in from these. (If all else fails a `peerDep` of `*` could do the trick, although I have not thought that through, so may have other issues.)

One thing that we could do is: we could publish an alpha version of `@substrate/connect` first which wouldn't have these dependencies (e.g: `@substrate/connect@7.0.0-alpha.0`), and then use that alpha version here and as soon as we release the stable version we upgraded it here with a patch. Please let me know if that's something that would work for you.

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts]: I've addressed all these ones already... I will address the ones on `Health` now.

**Comment by jacogr** [comment]
The beta works, didn't think of that, but it certainly the easiest.

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts]: All those global `eslint-disable` have now been removed :tada:

**Comment by tomaka** [comment]
> There are an important number of reasons why we decided to go with this API. Perhaps @tomaka is the best person for explaining why an API like new ScProvider(<chain info>) is not the best idea for a @substrate/connect Provider...

Well, an example problematic situation is:

- Library A depends on PolkadotJS, and calls `new ScProvider(polkadot)`
- Library B depends on PolkadotJS, and calls `new ScProvider(polkadot)`, except that their `polkadot` chain spec is slightly different
- Your project uses both library A and B, and so create two `ScProvider`s that are connected to two different chains. Smoldot cannot know which one of the two is the legitimate Polkadot. But so far everything works.
- Here is where the problem happens: if you then call `new ScProvider(somePolkadotParachain)`, smoldot has no idea which of the two polkadots to use for that parachain.

For this reason, adding a chain in the substrate-connect API is done in two steps: first, you create a client with `createScClient`, and then you add a chain using `client.addChain`. If you add the two polkadots to the same client then add a parachain, you get an error because of the ambiguity I've described. Instead, if you want to make it work, you're supposed to create a client with the first polkadot, a client with the second polkadot, and add the parachain to either of them.

In the example above, libraries A and B should each create their own "client" with `createScClient`, and then adding a parachain should be done explicitly to either of these clients.


So we could go for the `new ScProvider` API, and some kind of client as parameter, but if we do that then the implementation of `new ScProvider` needs to access the internals of the client, which means exposing the internals of the client.
That's why we went for this API with `createScClient` that returns an object with two methods.

**Comment by jacogr** [comment]
My complaint is rather just based on the consistency on _this_ layer, aka there is one way in which you create the exposed providers from `@polkadot/api` (despite the fact that their constructors are different based on the type of connection they do make). 

I'm guessing that if a client could be required in some cases, it can also be passed? e.g. 

```js
const relay = new ScProvider(relaySpec);

// optional client really needed if you do have the dual-lib issue described
// above (which would error otherwise)
const para = new ScProvider(paraSpec, relay.client); 
```

I don't have an answer of the specifics since I'm just really focussed on my own little island here.

**Comment by josepot** [comment]
> Yeah, we can indeed do that:
> 
> * `createScClient` stays as it is.
> * Add a new `ScProvider` whose constructor accepts as parameter either a chain spec or a well-known chain name, plus an optional client. If no client is provided it calls `createScClient`. In both cases, it calls `addChain` or `addWellKnownChain` on the client.

This is tricky, because both `addChain` and `addWellKnownChain` are asynchronous, while creating a new instance of a class is a synchronous call. So, even if we handle the tricky bits of that asynchronicity internally (which is a pain, but it's doable), the following code would still be problematic:

```js
const relay = new ScProvider(relaySpec);
const para = new ScProvider(paraSpec, relay.client); 
```

We can't just call `addChain`/`addWellKnownChain` for the parachain until the connection of the relayChain has resolved... I mean, we could handle that internally, but it's going to be a huge PITA and it's going to make it a lot more difficult for the user to handle the exceptions that can occur while the chain is being initialized.

**Review by jacogr — COMMENTED** [review]
Some final small comments to follow polkadot-js conventions with condition asserts. (One-line changes, import would be required from `@polkadot/util`)

Apart from the outstanding class-only interface, looks good as a first round.

**Review by jacogr — COMMENTED** [review]
Sorry, one hanging one added - GH was acting up a bit, so it got missed :(

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:238]: Not really, this assertion is bogus, because the previous line (`if(!this.isConnected) throw...`, or `assert(this.isConnected, 'Provider is not connected')`) already ensures that the `chain` is present. But it's up to you...

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:282]: Wouldn't you prefer?

```ts
assert(this.#currentSubscriptionId && !this.#currentSubunsubRequestId, 'Internal error in health checker'); 
```

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:238]: Resolving it b/c it's a duplicate

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:62]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:75]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:81]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:207]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:265]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:282]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:235]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:244]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:281]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:275]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:301]: Resolving it b/c it's a dup

File: packages/rpc-provider/src/substrate-connect/Health.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/Health.ts:75]: Shouldn't this be?

```ts
assert(checker === null, "Can't start the health checker multiple times in parallel");
```

Comment (jacogr) [reply packages/rpc-provider/src/substrate-connect/Health.ts:75]: 100%, I guess it shows it was pre coffee :)

File: packages/rpc-provider/src/substrate-connect/createScClient.ts
Comment (jacogr) [reply packages/rpc-provider/src/substrate-connect/createScClient.ts:244]: Sorry about that, I dunno wtf was wrong with GH this morning, for each it added 2 suggestions, even in the same comment box. (I had to edit those I missed afterwards)

**Comment by josepot** [comment]
Hi @jacogr ! think that I've now I've addressed all your comments. Could please re-review this one more time whenever you have a moment? Thanks a lot!

File: packages/rpc-provider/src/substrate-connect/ScProvider.ts
Comment (tomaka) [inline packages/rpc-provider/src/substrate-connect/ScProvider.ts:51]: This looks good but would rename `dependency` to `sharedSandbox`

**Review by jacogr — APPROVED** [review]
Nice! Ok, so next steps -

- just merge in master which should resolve the conflicts
- hopefully the above makes CI actually run ... let's not complain about GH actions lately or we will be here the whole day...
- once done, and green, will merge

After this let's get the alpha up and in here ASAP so we can remove the extra old deps here, otherwise that will create a bit of chaos on my releases over the weekend.

**Comment by josepot** [comment]
So, after having implemented the "class-only interface" in the way that @jacogr requested in [this comment](https://github.com/polkadot-js/api/pull/4672#issuecomment-1075032227). I've realized that if we make a small change, and we make the Provider not to auto-connect upon being instantiated (like it does by default on the WsProvider), then all my concerns about the user not being able to capture errors that could happen upon the initialization of the `@substrate/connect` chain go away.

I.e what I'm suggesting is that the correct usage of the ScProvider would be:

```ts
import { ScProvider } from '@polkadot/rpc-provider';
const polkadotProvider = new ScProvider(SupportedChains.polkadot);
const parachainProvider = new ScProvider(parachainSpec, polkadotProvider);
await parachainProvider.connect()
const version = await parachainProvider.send('chain_getBlockHash', []);
```

I'm going to go ahead and implement this last change. I hope that we all agree that this is best :pray:.

File: packages/rpc-provider/src/substrate-connect/ScProvider.ts
Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/ScProvider.ts:51]: I think that I understand where this is coming from. However, from the point of view of the consumer, this is a dependency. As in the connection of this Provider that I'm creating depends on the connection of this other Provider... The fact that they have a shared sandbox behind the scenes it's a very important implementation detail, but I'm of the opinion that the end user doesn't have to know about it. Although, in the end this should be your call, so if you disagree with this reasoning I will rename it to `sharedSandbox`, for sure!

Comment (tomaka) [reply packages/rpc-provider/src/substrate-connect/ScProvider.ts:51]: Please rename to `sharedSandbox`, it's not just a parachain <-> relay chain thing. Sorry for the lack of deep explanation but I've explained this at least twenty times to many different people.

Comment (josepot) [reply packages/rpc-provider/src/substrate-connect/ScProvider.ts]: done!

File: packages/rpc-provider/package.json
Comment (jacogr) [inline packages/rpc-provider/package.json:35]: Thank you.

Comment (josepot) [reply packages/rpc-provider/package.json:35]: Totally! It was the right thing to do! So, thank you for pointing it out!

**Comment by jacogr** [comment]
Could you just push a commit to make GH actions wake up? (According to their status page the maintenance is now completed... so let's hope our lives improve...)

You can just increment the first number in the `.123trigger` file in the root, will be enough.

**Comment by josepot** [comment]
> Could you just push a commit to make GH actions wake up? (According to their status page the maintenance is now completed and actions are supposed to be humming again... so let's hope our lives improve...)
> 
> You can just increment the first number in the `.123trigger` file in the root, will be enough.

Yep! I just merged master for triggering the CI... I hope it works :crossed_fingers:

**Pull request merged**
